File: test/aidbox_sdk/snapshots/dotnethl7-fhir-r4-core/PlanDefinition.cs
using Aidbox.FHIR.Base;
using Aidbox.FHIR.Utils;

namespace Aidbox.FHIR.R4.Core;

public class PlanDefinition : DomainResource, IResource
{
    public string? Description { get; set; }
    public string? Date { get; set; }
    public Base.ContactDetail[]? Endorser { get; set; }
    public string? Publisher { get; set; }
    public string? ApprovalDate { get; set; }
    public Base.CodeableConcept[]? Jurisdiction { get; set; }
    public string? Purpose { get; set; }
    public Base.CodeableConcept? SubjectCodeableConcept { get; set; }
    public string? Name { get; set; }
    public Base.UsageContext[]? UseContext { get; set; }
    public PlanDefinition_Goal[]? Goal { get; set; }
    public string? Copyright { get; set; }
    public Base.CodeableConcept? Type { get; set; }
    public bool? Experimental { get; set; }
    public Base.CodeableConcept[]? Topic { get; set; }
    public string? Title { get; set; }
    public string[]? Library { get; set; }
    public Base.ContactDetail[]? Author { get; set; }
    public string? Usage { get; set; }
    public required string Status { get; set; }
    public string? Subtitle { get; set; }
    public string? Url { get; set; }
    public Base.Identifier[]? Identifier { get; set; }
    public string? LastReviewDate { get; set; }
    public Base.ContactDetail[]? Editor { get; set; }
    public PlanDefinition_Action[]? Action { get; set; }
    public Base.ContactDetail[]? Reviewer { get; set; }
    public string? Version { get; set; }
    public Base.RelatedArtifact[]? RelatedArtifact { get; set; }
    public Base.ContactDetail[]? Contact { get; set; }
    public Base.ResourceReference? SubjectReference { get; set; }
    public object? Subject    
    {
        get
        {
            if (SubjectCodeableConcept is not null)
            {
                return SubjectCodeableConcept;
            }
    
            if (SubjectReference is not null)
            {
                return SubjectReference;
            }
    
            return null;
        }
    
        set
        {
            if (value?.GetType() == typeof(Base.CodeableConcept))
            {
                SubjectCodeableConcept = (Base.CodeableConcept)value;
                return;
            }
    
            if (value?.GetType() == typeof(Base.ResourceReference))
            {
                SubjectReference = (Base.ResourceReference)value;
                return;
            }
    
            throw new ArgumentException("Invalid type provided");
        }
    }
    public Base.Period? EffectivePeriod { get; set; }

    public class PlanDefinition_Goal_Target : BackboneElement
    {
        public string? Due { get; set; }
        public object? Detail    
        {
            get
            {
                if (DetailRange is not null)
                {
                    return DetailRange;
                }
        
                if (DetailQuantity is not null)
                {
                    return DetailQuantity;
                }
        
                if (DetailCodeableConcept is not null)
                {
                    return DetailCodeableConcept;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.Range))
                {
                    DetailRange = (Base.Range)value;
                    return;
                }
        
                if (value?.GetType() == typeof(Base.Quantity))
                {
                    DetailQuantity = (Base.Quantity)value;
                    return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    DetailCodeableConcept = (Base.CodeableConcept)value;
                    return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.CodeableConcept? Measure { get; set; }
        public Base.Range? DetailRange { get; set; }
        public Base.Quantity? DetailQuantity { get; set; }
        public Base.CodeableConcept? DetailCodeableConcept { get; set; }
    }

    public class PlanDefinition_Goal : BackboneElement
    {
        public Base.CodeableConcept? Start { get; set; }
        public PlanDefinition_Goal_Target[]? Target { get; set; }
        public Base.CodeableConcept? Category { get; set; }
        public Base.CodeableConcept? Priority { get; set; }
        public Base.CodeableConcept[]? Addresses { get; set; }
        public required Base.CodeableConcept Description { get; set; }
        public Base.RelatedArtifact[]? Documentation { get; set; }
    }

    public class PlanDefinition_Action_RelatedAction : BackboneElement
    {
        public object? Offset    
        {
            get
            {
                if (OffsetRange is not null)
                {
                    return OffsetRange;
                }
        
                if (OffsetDuration is not null)
                {
                    return OffsetDuration;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.Range))
                {
                    OffsetRange = (Base.Range)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    OffsetDuration = (string)value;
                    return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public required string ActionId { get; set; }
        public Base.Range? OffsetRange { get; set; }
        public required string Relationship { get; set; }
        public string? OffsetDuration { get; set; }
    }

    public class PlanDefinition_Action_Participant : BackboneElement
    {
        public Base.CodeableConcept? Role { get; set; }
        public required string Type { get; set; }
    }

    public class PlanDefinition_Action_Condition : BackboneElement
    {
        public required string Kind { get; set; }
        public Base.ResourceExpression? Expression { get; set; }
    }

    public class PlanDefinition_Action_DynamicValue : BackboneElement
    {
        public string? Path { get; set; }
        public Base.ResourceExpression? Expression { get; set; }
    }

    public class PlanDefinition_Action : BackboneElement
    {
        public Base.Range? TimingRange { get; set; }
        public string? Description { get; set; }
        public object? Definition    
        {
            get
            {
                if (DefinitionUri is not null)
                {
                    return DefinitionUri;
                }
        
                if (DefinitionCanonical is not null)
                {
                    return DefinitionCanonical;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(string))
                {
                    DefinitionUri = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    DefinitionCanonical = (string)value;
                    return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public string? Transform { get; set; }
        public string? TextEquivalent { get; set; }
        public string? DefinitionUri { get; set; }
        public string[]? GoalId { get; set; }
        public Base.CodeableConcept? SubjectCodeableConcept { get; set; }
        public Base.Period? TimingPeriod { get; set; }
        public string? DefinitionCanonical { get; set; }
        public PlanDefinition_Action_RelatedAction[]? RelatedAction { get; set; }
        public Base.CodeableConcept? Type { get; set; }
        public PlanDefinition_Action_Participant[]? Participant { get; set; }
        public Base.DataRequirement[]? Output { get; set; }
        public string? Title { get; set; }
        public Base.RelatedArtifact[]? Documentation { get; set; }
        public string? Prefix { get; set; }
        public string? SelectionBehavior { get; set; }
        public Base.CodeableConcept[]? Reason { get; set; }
        public string? TimingDateTime { get; set; }
        public string? TimingTiming { get; set; }
        public string? TimingDuration { get; set; }
        public string? Priority { get; set; }
        public string? RequiredBehavior { get; set; }
        public PlanDefinition_Action_Condition[]? Condition { get; set; }
        public object? Timing    
        {
            get
            {
                if (TimingRange is not null)
                {
                    return TimingRange;
                }
        
                if (TimingPeriod is not null)
                {
                    return TimingPeriod;
                }
        
                if (TimingDateTime is not null)
                {
                    return TimingDateTime;
                }
        
                if (TimingTiming is not null)
                {
                    return TimingTiming;
                }
        
                if (TimingDuration is not null)
                {
                    return TimingDuration;
                }
        
                if (TimingAge is not null)
                {
                    return TimingAge;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.Range))
                {
                    TimingRange = (Base.Range)value;
                    return;
                }
        
                if (value?.GetType() == typeof(Base.Period))
                {
                    TimingPeriod = (Base.Period)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    TimingDateTime = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    TimingTiming = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    TimingDuration = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    TimingAge = (string)value;
                    return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public string? GroupingBehavior { get; set; }
        public PlanDefinition_Action_DynamicValue[]? DynamicValue { get; set; }
        public Base.CodeableConcept[]? Code { get; set; }
        public string? TimingAge { get; set; }
        public Base.ResourceReference[]? Action { get; set; }
        public string? PrecheckBehavior { get; set; }
        public Base.DataRequirement[]? Input { get; set; }
        public Base.TriggerDefinition[]? Trigger { get; set; }
        public Base.ResourceReference? SubjectReference { get; set; }
        public object? Subject    
        {
            get
            {
                if (SubjectCodeableConcept is not null)
                {
                    return SubjectCodeableConcept;
                }
        
                if (SubjectReference is not null)
                {
                    return SubjectReference;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    SubjectCodeableConcept = (Base.CodeableConcept)value;
                    return;
                }
        
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    SubjectReference = (Base.ResourceReference)value;
                    return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public string? CardinalityBehavior { get; set; }
    }
}

File: test/aidbox_sdk/snapshots/python/hl7-fhir-r4-core/PlanDefinition.py
from typing import Optional, List
from pydantic import *
from ..base import *

class PlanDefinition_Goal_Target(BackboneElement):
    due: Optional[Duration] = None
    measure: Optional[CodeableConcept] = None
    detail_range: Optional[Range] = None
    detail_quantity: Optional[Quantity] = None
    detail_codeable_concept: Optional[CodeableConcept] = None

class PlanDefinition_Goal(BackboneElement):
    start: Optional[CodeableConcept] = None
    target: Optional[List[PlanDefinition_Goal_Target]] = None
    category: Optional[CodeableConcept] = None
    priority: Optional[CodeableConcept] = None
    addresses: Optional[List[CodeableConcept]] = None
    description: CodeableConcept
    documentation: Optional[List[RelatedArtifact]] = None

class PlanDefinition_Action_RelatedAction(BackboneElement):
    action_id: str
    offset_range: Optional[Range] = None
    relationship: str
    offset_duration: Optional[Duration] = None

class PlanDefinition_Action_Participant(BackboneElement):
    role: Optional[CodeableConcept] = None
    type: str

class PlanDefinition_Action_Condition(BackboneElement):
    kind: str
    expression: Optional[Expression] = None

class PlanDefinition_Action_DynamicValue(BackboneElement):
    path: Optional[str] = None
    expression: Optional[Expression] = None

class PlanDefinition_Action(BackboneElement):
    timing_range: Optional[Range] = None
    description: Optional[str] = None
    transform: Optional[str] = None
    text_equivalent: Optional[str] = None
    definition_uri: Optional[str] = None
    goal_id: Optional[List[str]] = None
    subject_codeable_concept: Optional[CodeableConcept] = None
    timing_period: Optional[Period] = None
    definition_canonical: Optional[str] = None
    related_action: Optional[List[PlanDefinition_Action_RelatedAction]] = None
    type: Optional[CodeableConcept] = None
    participant: Optional[List[PlanDefinition_Action_Participant]] = None
    output: Optional[List[DataRequirement]] = None
    title: Optional[str] = None
    documentation: Optional[List[RelatedArtifact]] = None
    prefix: Optional[str] = None
    selection_behavior: Optional[str] = None
    reason: Optional[List[CodeableConcept]] = None
    timing_date_time: Optional[str] = None
    timing_timing: Optional[Timing] = None
    timing_duration: Optional[Duration] = None
    priority: Optional[str] = None
    required_behavior: Optional[str] = None
    condition: Optional[List[PlanDefinition_Action_Condition]] = None
    grouping_behavior: Optional[str] = None
    dynamic_value: Optional[List[PlanDefinition_Action_DynamicValue]] = None
    code: Optional[List[CodeableConcept]] = None
    timing_age: Optional[Age] = None
    action: Optional[List[Reference]] = None
    precheck_behavior: Optional[str] = None
    input: Optional[List[DataRequirement]] = None
    trigger: Optional[List[TriggerDefinition]] = None
    subject_reference: Optional[Reference] = None
    cardinality_behavior: Optional[str] = None

class PlanDefinition(DomainResource):
    description: Optional[str] = None
    date: Optional[str] = None
    endorser: Optional[List[ContactDetail]] = None
    publisher: Optional[str] = None
    approval_date: Optional[str] = None
    jurisdiction: Optional[List[CodeableConcept]] = None
    purpose: Optional[str] = None
    subject_codeable_concept: Optional[CodeableConcept] = None
    name: Optional[str] = None
    use_context: Optional[List[UsageContext]] = None
    goal: Optional[List[PlanDefinition_Goal]] = None
    copyright: Optional[str] = None
    type: Optional[CodeableConcept] = None
    experimental: Optional[bool] = None
    topic: Optional[List[CodeableConcept]] = None
    title: Optional[str] = None
    library: Optional[List[str]] = None
    author: Optional[List[ContactDetail]] = None
    usage: Optional[str] = None
    status: str
    subtitle: Optional[str] = None
    url: Optional[str] = None
    identifier: Optional[List[Identifier]] = None
    last_review_date: Optional[str] = None
    editor: Optional[List[ContactDetail]] = None
    action: Optional[List[PlanDefinition_Action]] = None
    reviewer: Optional[List[ContactDetail]] = None
    version: Optional[str] = None
    related_artifact: Optional[List[RelatedArtifact]] = None
    contact: Optional[List[ContactDetail]] = None
    subject_reference: Optional[Reference] = None
    effective_period: Optional[Period] = None